### PR TITLE
Add address field

### PR DESCRIPTION
**Pull request by IgorZiegler**
- add address block which use typeahead autocomplete
 - add example with address block

### CODE REVIEW

File: frontend/blocks/address/address.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import Fieldset from '../fieldset/fieldset';
+import Factory from '../factory';
+import './address.scss';
+
+import Bloodhound from 'typeahead.js';
+import defaults from 'lodash/defaults';
+
+export default class Address extends Fieldset {
+  constructor(config) {
+    const finalConfig = defaults(config, {
```
Comment (abukurov) [inline frontend/blocks/address/address.js]: Config object should be immutable

File: frontend/blocks/address/address.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  setAddressPartValue(fieldName, value) {
+    const field = this.getItemByName(fieldName);
+
+    if (!field) {
+      return;
+    }
+    field.value = value;
+  }
+
+  pushAddressPart(streetParts, addressPartValue) {
```
Comment (maxmalov) [inline frontend/blocks/address/address.js]: why is this a method?

File: frontend/blocks/address/address.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+          const ajax = Object.assign({
+            headers: {}
+          }, me.config.ajax);
+
+          settings.type = ajax && ajax.type || 'POST';
+          settings.dataType = ajax && ajax.dataType || 'json';
+          settings.headers = settings.headers || {};
+          settings.headers.Authorization = 'Token ' + (ajax && ajax.headers.dataApiKey);
+          settings.headers['Content-Type'] = ajax && ajax.headers.contentType || 'application/json';
+          settings.data = JSON.stringify(defaults(me.config.params, {
```
Comment (abukurov) [inline frontend/blocks/address/address.js]: You can omit `JSON.stringify`

File: frontend/blocks/address/address.js (hunk 2 above)
Comment (maxmalov) [inline frontend/blocks/address/address.js]: why it mutate input arguments?

File: frontend/blocks/address/address.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+          name: 'apartment'
+        }
+      ]
+    });
+
+    super(finalConfig);
+  }
+
+  afterRender() {
+    const me = this;
```
Comment (abukurov) [inline frontend/blocks/address/address.js]: ???

File: frontend/blocks/address/address.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+          settings.dataType = ajax && ajax.dataType || 'json';
+          settings.headers = settings.headers || {};
+          settings.headers.Authorization = 'Token ' + (ajax && ajax.headers.dataApiKey);
+          settings.headers['Content-Type'] = ajax && ajax.headers.contentType || 'application/json';
+          settings.data = JSON.stringify(defaults(me.config.params, {
+            query: query || '',
+            count: 20,
+            locations: me.config.regions.map(region => ({region}))
+          }));
+          settings.url = settings.url + '?regions=' + me.config.regions.join(',') + '&q=' + query;
```
Comment (abukurov) [inline frontend/blocks/address/address.js]: string templates

File: frontend/blocks/fieldset/fieldset.js [hunk 6]
```diff
@@ -69,6 +69,18 @@ export default class Fieldset extends Container {
   appendChild(block) {
     this.el.find('.fieldset-rows-container').append(block.el);
   }
+
+  getItemByName(name) {
+    let item;
```
Comment (abukurov) [inline frontend/blocks/fieldset/fieldset.js]: image

File: frontend/blocks/address/address.scss [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+$cursorBackgroundColor: #428bca;
+$cursorTextColor: #fff;
+
+.#{$cls-prefix}-fieldset {
+  .twitter-typeahead {
+    width: 100%;
+  }
+
+  .tt-menu [class*=tt-dataset]{
+    max-height: inherit;
```
Comment (102) [inline frontend/blocks/address/address.scss]: this should be inside `.tt-menu` block

File: frontend/blocks/address/address.scss [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+$borderColor: #cccccc;
+$cursorBackgroundColor: #428bca;
+$cursorTextColor: #fff;
+
+.#{$cls-prefix}-fieldset {
+  .twitter-typeahead {
+    width: 100%;
+  }
+
+  .tt-menu [class*=tt-dataset]{
```
Comment (102) [inline frontend/blocks/address/address.scss]: this should be inside .tt-menu block

File: frontend/blocks/address/address.scss [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  .tt-suggestion {
+    display: block;
+    padding: 3px 20px;
+    &.tt-cursor, &:hover {
+      color: $cursorTextColor;
+      background-color: $cursorBackgroundColor;
+      a: {
+        color: $cursorTextColor;
+      }
+    };
```
Comment (102) [inline frontend/blocks/address/address.scss]: `;`?

File: frontend/blocks/address/address.scss [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    z-index: 900 !important;
+  }
+
+  .tt-suggestion {
+    display: block;
+    padding: 3px 20px;
+    &.tt-cursor, &:hover {
+      color: $cursorTextColor;
+      background-color: $cursorBackgroundColor;
+      a: {
```
Comment (102) [inline frontend/blocks/address/address.scss]: Why do you need `colon` after `a`?

File: public/examples/forms/address.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
+{
+  "cls": "form",
+  "items": [
+    {
+      "block": "address",
+      "minLength": 3,
+      "highlight": true,
+      "label": "Адрес",
+      "layout": [ 1, 3],
```
Comment (102) [inline public/examples/forms/address.json]: extra space after `[`

File: public/examples/forms/address.json [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      "name": "Адрес",
+      "params": {
+        "query": "",
+        "count": 20
+      },
+      "regions": ["Омская"],
+      "url": "https://dadata.ru/api/v2/suggest/address/",
+      "ajax": {
+        "headers": {
+          "dataApiKey":"ffc44375a4d8268790b8aa1bc9d840aedd5aca40"
```
Comment (102) [inline public/examples/forms/address.json]: missing space symbol after `:`

File: frontend/blocks/address/address.js (hunk 1 above)
Comment (mvahrameev) [inline frontend/blocks/address/address.js]: ```defaults({...}, config)```

File: frontend/blocks/address/address.js [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    addressAutocompleteInput
+      .typeahead({
+        minLength: this.config.autocompleteMinLength || 2,
+        highlight: this.config.highlight || true
+      }, {
+        name: 'address',
+        displayKey: 'value',
+        source: address.ttAdapter()
+      })
+      .on('typeahead:selected', (obj, selected) => {
```
Comment (mvahrameev) [inline frontend/blocks/address/address.js]: extract function

**Pull request closed without merging**